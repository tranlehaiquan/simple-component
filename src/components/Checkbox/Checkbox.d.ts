import * as React from 'react';

export interface CheckboxProps {
  children?: React.ReactNode,
  checked?: boolean,
  onChange? (event: React.SyntheticEvent): void,
  value: [number | string | boolean],
  name?: string,
  disabled?: boolean,
  className?: string,
  required?: boolean,
  indeterminate?: boolean
}

declare const Checkbox: React.ComponentType<CheckboxProps>;

export default Checkbox;
