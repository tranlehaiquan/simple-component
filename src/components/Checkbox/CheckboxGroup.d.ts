import * as React from 'react';

export interface CheckboxGroupProps {
  children?: React.ReactNode,
  checked?: boolean,
  onChange? (event: React.SyntheticEvent): void,
  values: Array<number | string | boolean>,
  name?: string,
  disabled?: boolean,
  className?: string,
  required?: boolean,
  indeterminate?: boolean,
  min?: number,
  max?: number
}

declare const CheckboxGroup: React.ComponentType<CheckboxGroupProps>;

export default CheckboxGroup;
