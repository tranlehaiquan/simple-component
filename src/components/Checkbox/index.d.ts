export { default as CheckboxGroup } from './CheckboxGroup';
export { default } from './Checkbox';
