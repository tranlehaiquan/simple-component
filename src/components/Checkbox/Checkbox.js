import React, { memo } from 'react';
import classnames from 'classnames';
import propTypes from 'prop-types';

function Checkbox(props) {
  function handleOnChange(event) {
    const { onChange, value } = props;

    onChange(value, event);
  }

  function handleKeyPress(e) {
    const keyCodeOfSpace = 32;

    if (e.keyCode !== keyCodeOfSpace) return;

    const { onChange, value } = props;
    onChange(value);
  }

  const { children, checked, value, name, disabled, className, required, indeterminate } = props;
  const classnameCheckbox = classnames('sp-checkbox__item', {
    ['sp-checkbox__item--checked']: checked,
    ['sp-checkbox__item--disabled']: disabled,
    ['sp-checkbox__item--indeterminate']: indeterminate
  }, className);

  return (
    <label className={classnameCheckbox}>
      <input
        type="checkbox"
        className="sp-checkbox__input"
        name={name}
        value={value}
        tabIndex="-1"
        checked={checked}
        onChange={handleOnChange}
        disabled={disabled}
        required={required}
      />
      <span
        className="sp-checkbox__icon"
        tabIndex={disabled ? '-1' : '0'}
        onKeyDown={handleKeyPress}
      />
      <span className="sp-checkbox__text">{children || value}</span>
    </label>
  );
}

Checkbox.propTypes = {
  children: propTypes.node,
  checked: propTypes.bool,
  onChange: propTypes.func,
  value: propTypes.oneOfType([
    propTypes.number,
    propTypes.string,
    propTypes.bool,
  ]),
  name: propTypes.string,
  disabled: propTypes.bool,
  className: propTypes.string,
  required: propTypes.bool,
  indeterminate: propTypes.bool
};

Checkbox.defaultProps = {
  checked: false,
  disabled: false,
  required: false,
  name: '',
  className: ''
};

export default memo(Checkbox);
