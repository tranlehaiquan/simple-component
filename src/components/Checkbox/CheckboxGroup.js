import React, {
  Component
} from 'react';
import propTypes from 'prop-types';

class CheckboxGroup extends Component {
  handleOnClick = (value) => {
    this.props.onChange(value);
  }

  renderChildrent = () => {
    const {
      children,
      value,
      name
    } = this.props;

    const radios = React.Children.map(children, (radio) => {
      const checked = value.includes(radio.props.value);

      return React.cloneElement(
        radio, {
          checked,
          onChange: this.handleOnClick,
          name
        }
      );
    });

    return radios;
  }

  render() {
    return ( 
      <div className = "sp-radio__group" > 
        {
          this.renderChildrent()
        } 
      </div>
    );
  }
}

CheckboxGroup.propTypes = {
  children: propTypes.node,
  value: propTypes.array,
  name: propTypes.string,
  onChange: propTypes.func,
};

CheckboxGroup.defaultProps = {
  name: '',
  value: []
};

export default CheckboxGroup;
