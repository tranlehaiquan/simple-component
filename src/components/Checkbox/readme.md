# Checkbox, CheckboxGroup description

## Checkbox

Checkbox element can be use with or without CheckboxGroup

### Todo

- [x] Add indeterminate `/ˌɪn.dɪˈtɝː.mɪ.nət/` for checkbox
- [ ] Add min/ max for checkboxGroup

### Basic usage

```Javascript
  <Checkbox value="organe" checked={true} onChange={onChange}/>
  <Checkbox value="banana" checked={false} onChange={onChange}/>
  <Checkbox value="coconut" checked={false} onChange={onChange}/>
```

### Disabled usage

```Javascript
  <Checkbox value="organe" disabled checked={true} onChange={onChange}/>
  <Checkbox value="banana" disabled checked={false} onChange={onChange}/>
  <Checkbox value="coconut" disabled checked={false} onChange={onChange}/>
```

## Checkbox group

CheckboxGroup component used for multiple Checkbox, indicates with Checkbox is checked

### Basic

```Javascript
  <CheckboxGroup value="organe" onChange={onChange}>
    <Checkbox value="organe" />
    <Checkbox value="banana" />
    <Checkbox value="coconut" />
  </CheckboxGroup>
```

### Disabled usage

Usage for disable all checkbox inside

```Javascript
  <CheckboxGroup value="organe" onChange={onChange} disabled>
    <Checkbox value="organe" />
    <Checkbox value="banana" />
    <Checkbox value="coconut" />
  </CheckboxGroup>
```

### Checkbox event

| Event Name | Description                                    | Parameters   |
|------------|------------------------------------------------|--------------|
| onChange   | This event apply for Checkbox or CheckboxGroup | Value, Event |

### Checkbox attribute

| Prop           | Description                | Type                  | Options | Default |
|----------------|----------------------------|-----------------------|---------|---------|
| children       | Enable user custom present | Node                  |         |         |
| checked        | Whether checked or not     | Boolean               |         | false   |
| onChange       | Event on change            | Function              |         |         |
| value          |                            | Number/String/Boolean |         |         |
| name           |                            | String                |         | ''      |
| disabled       |                            | Boolean               |         | false   |
| className      |                            | String                |         | ''      |
| required       |                            | Boolean               |         | false   |
| indeterminate  |                            | Boolean               |         | false   |

### CheckboxGroup attribute

| Prop      | Description                             | Type                  | Options | Default |
|-----------|-----------------------------------------|-----------------------|---------|---------|
| children  | Checkbox components                     | Node                  |         |         |
| onChange  | Event on change                         | Function              |         |         |
| value     | For current value, not Checkbox itselft | Number/String/Boolean |         |         |
| name      |                                         | String                |         | ''      |
| disabled  |                                         | Boolean               |         | false   |
| className |                                         | String                |         | ''      |
