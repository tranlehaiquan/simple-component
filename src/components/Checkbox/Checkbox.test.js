import React from 'react';
import renderer from 'react-test-renderer';
import Checkbox from './Checkbox';
import { shallow } from 'enzyme';
import sinon from 'sinon';

describe('snapshot for checkbox', () => {
  test('Checkbox', () => {
    const component = renderer.create(
      <div>
        <Checkbox value="organe" name="vegetable" checked={true} onChange={() => {}}/>
        <Checkbox value="banana" name="vegetable" checked={false} disabled onChange={() => {}}/>
        <Checkbox value="coconut" name="vegetable" checked={false} required onChange={() => {}}/>
        <Checkbox value="coconut" name="vegetable" checked={false} onChange={() => {}} className="sp-last"/>
      </div>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Checkbox behavior', () => {
  test('Checkbox without group, spy', () => {
    const onChange = sinon.spy();
    const component = shallow(
      <div>
        <Checkbox value="organe" checked={true} onChange={onChange}/>
        <Checkbox value="banana" checked={false} onChange={onChange}/>
        <Checkbox value="coconut" checked={false} onChange={onChange}/>
      </div>
    );

    const checkbox = component.find('Checkbox').first();
    checkbox.simulate('change');

    expect(onChange).toHaveProperty('callCount', 1);
  });
});
