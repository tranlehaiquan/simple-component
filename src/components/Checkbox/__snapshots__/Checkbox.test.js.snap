// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for checkbox Checkbox 1`] = `
<label
  className="sp-checkbox__item"
>
  <input
    checked={false}
    className="sp-checkbox__input"
    disabled={false}
    onChange={[Function]}
    tabIndex="-1"
    type="checkbox"
    value="Hi"
  />
  <span
    className="sp-checkbox__icon"
    onKeyDown={[Function]}
    tabIndex="0"
  />
  <span
    className="sp-checkbox__text"
  >
    Hi
  </span>
</label>
`;

exports[`snapshot for checkbox Checkbox 2`] = `
<div
  className="sp-radio__group"
>
  <label
    className="sp-checkbox__item"
  >
    <input
      checked={false}
      className="sp-checkbox__input"
      disabled={false}
      name=""
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="organe"
    />
    <span
      className="sp-checkbox__icon"
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <span
      className="sp-checkbox__text"
    >
      organe
    </span>
  </label>
  <label
    className="sp-checkbox__item sp-checkbox__item--checked"
  >
    <input
      checked={true}
      className="sp-checkbox__input"
      disabled={false}
      name=""
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="banana"
    />
    <span
      className="sp-checkbox__icon"
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <span
      className="sp-checkbox__text"
    >
      banana
    </span>
  </label>
  <label
    className="sp-checkbox__item"
  >
    <input
      checked={false}
      className="sp-checkbox__input"
      disabled={false}
      name=""
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="coconut"
    />
    <span
      className="sp-checkbox__icon"
      onKeyDown={[Function]}
      tabIndex="0"
    />
    <span
      className="sp-checkbox__text"
    >
      coconut
    </span>
  </label>
  <label
    className="sp-checkbox__item sp-checkbox__item--disabled"
  >
    <input
      checked={false}
      className="sp-checkbox__input"
      disabled={true}
      name=""
      onChange={[Function]}
      tabIndex="-1"
      type="checkbox"
      value="carot"
    />
    <span
      className="sp-checkbox__icon"
      onKeyDown={[Function]}
      tabIndex="-1"
    />
    <span
      className="sp-checkbox__text"
    >
      carot
    </span>
  </label>
</div>
`;
